### ISSUE TITLE
building the library for windows

### ISSUE THREAD

**Issue by abdelrahman1215** [libconfuse/libconfuse#172]
I ran configure with the --prefix=mingw flag and then when i ran make i got this linker error
`make  all-recursive
make[1]: Entering directory '/c/Users/abder/Downloads/confuse-3.3'
Making all in m4
make[2]: Entering directory '/c/Users/abder/Downloads/confuse-3.3/m4'
make[2]: Nothing to be done for 'all'.
make[2]: Leaving directory '/c/Users/abder/Downloads/confuse-3.3/m4'
Making all in po
make[2]: Entering directory '/c/Users/abder/Downloads/confuse-3.3/po'
make[2]: Nothing to be done for 'all'.
make[2]: Leaving directory '/c/Users/abder/Downloads/confuse-3.3/po'
Making all in src
make[2]: Entering directory '/c/Users/abder/Downloads/confuse-3.3/src'
/bin/sh ../libtool  --tag=CC   --mode=link gcc  -g -O2  -no-undefined -version-info 3:0:1  -o libconfuse.la -rpath /mingw/lib libconfuse_la-confuse.lo libconfuse_la-lexer.lo libconfuse_la-fmemopen.lo libconfuse_la-reallocarray.lo -lintl
libtool: link: rm -fr  .libs/libconfuse.dll.a
libtool: link: gcc -shared  .libs/libconfuse_la-confuse.o   -lintl  -g -O2   -o .libs/libconfuse-2.dll -Wl,--enable-auto-image-base -Xlinker --out-implib -Xlinker .libs/libconfuse.dll.a
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: .libs/libconfuse_la-confuse.o: in function `cfg_free':
C:/Users/abder/Downloads/confuse-3.3/src/confuse.c:1992: undefined reference to `cfg_yylex_destroy'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: .libs/libconfuse_la-confuse.o: in function `cfg_init_defaults':
C:/Users/abder/Downloads/confuse-3.3/src/confuse.c:792: undefined reference to `fmemopen'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:/Users/abder/Downloads/confuse-3.3/src/confuse.c:801: undefined reference to `cfg_scan_fp_begin'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:/Users/abder/Downloads/confuse-3.3/src/confuse.c:808: undefined reference to `cfg_scan_fp_end'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: .libs/libconfuse_la-confuse.o: in function `cfg_parse_internal':
C:/Users/abder/Downloads/confuse-3.3/src/confuse.c:1287: undefined reference to `cfg_yylex'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: .libs/libconfuse_la-confuse.o: in function `cfg_addopt':
C:/Users/abder/Downloads/confuse-3.3/src/confuse.c:627: undefined reference to `reallocarray'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: .libs/libconfuse_la-confuse.o: in function `cfg_parse_fp':
C:/Users/abder/Downloads/confuse-3.3/src/confuse.c:1677: undefined reference to `cfg_scan_fp_begin'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:/Users/abder/Downloads/confuse-3.3/src/confuse.c:1679: undefined reference to `cfg_scan_fp_end'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: .libs/libconfuse_la-confuse.o: in function `cfg_parse_buf':
C:/Users/abder/Downloads/confuse-3.3/src/confuse.c:1798: undefined reference to `fmemopen'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: .libs/libconfuse_la-confuse.o: in function `cfg_include':
C:/Users/abder/Downloads/confuse-3.3/src/confuse.c:2011: undefined reference to `cfg_lexer_include'
collect2.exe: error: ld returned 1 exit status
make[2]: *** [Makefile:452: libconfuse.la] Error 1
make[2]: Leaving directory '/c/Users/abder/Downloads/confuse-3.3/src'
make[1]: *** [Makefile:520: all-recursive] Error 1
make[1]: Leaving directory '/c/Users/abder/Downloads/confuse-3.3'
make: *** [Makefile:406: all] Error 2`

**Comment by troglobit** [comment]
There's a `windows/mingw/README.TXT` did you see that?

**Comment by abdelrahman1215** [comment]
yes and i followed the instructions but when i tried to debug a simple
program to test it gdb kept exiting and when i ran it on bash it said that
it needs libconfuse-2.dll so i tried to build it but i got the linker
message that mentioned in the issue

**Comment by abdelrahman1215** [comment]
I finally solved the proplem
so this is what i had to do if any one else encountered the same proplem :

1. I moved the definition of HAVE_WINDOWS_H from config.h to fmemopen.c.
2. I included reallocarray.c in confuse.c.

**Comment by troglobit** [comment]
OK, but none of that should be necessary:

 - config.h should be included in fmemopen.c when HAVE_CONFIG_H is set on the command-line, which is done on all other build targets
 - reallocarray.c is built-in to libconfuse for systems where the configure script cannot find reallocarray() in the libc

Exactly how did you autogen.sh + configure + make libconfuse to get to this state? And what version of libconfuse are you using, or are you building directly from the latest master branch on GitHub?

**Comment by abdelrahman1215** [comment]
I downloaded the zip file from the releases page yesterday so i think it is the latest release and i configured and built it using
 mingw64 bash i tried to build it multiple times but here is what i did the last time

1. I decompressed the files agian in case it was corrrupted
2. I ran configure with no arguments
3. I ran make to build the dll

when i opened the fmemopen.c to see what is the proplem config.h was included and it had HAVE_WINDOWS_H set to 1 but for some reason all the definitions of fmemopen where dark which means that it is not activated due to preprocessors so i moved HAVE_WINDOWS_H from config.h to fmemopen and when i ran make the only error i got was a linker error saying that it can't find the definition of realloc_array so i included it in confuse.c and finally i was able to build the dll

**Comment by troglobit** [comment]
Thanks for detailing your answer.  We'll have a look at it later to see if we can figure out what's wrong.

**Comment by AndreiCherniaev** [comment]
@abdelrahman1215 can you try [conan way](https://github.com/libconfuse/libconfuse/pull/169)? 

P.S. I still believe that it should be officially supported by libconfuse...
